chain-storage/src/lib.rs

```diff
@@ -1,1079 +1,827 @@
-                blocks_fetched += 1;
-            })
-            .unwrap();
-
-            assert_eq!(
-                ancestor_info.chain_length + distance,
-                block.chain_length().0
-            );
+    let block_info_bin = info.get(block_id)?.ok_or(ConsistencyFailure::BlockInfo)?;
+    let mut block_info_reader: &[u8] = &block_info_bin;
+    let block_info = BlockInfo::deserialize(&mut block_info_reader, id_size, block_id.to_vec())?;
 
-            let ancestor = store.get_block(&ancestor_info.block_hash).unwrap().0;
+    if block_info.ref_count() != 0 {
+        return Ok(RemoveTipResult::Done);
+    }
 
-            assert_eq!(ancestor.chain_length().0 + distance, block.chain_length().0);
-        }
+    info.remove(block_id)?;
+    blocks.remove(block_id)?;
 
-        let blocks_per_test = blocks_fetched as f64 / nr_tests as f64;
+    let mut chain_length_index = block_info.chain_length().to_le_bytes().to_vec();
+    chain_length_index.extend_from_slice(block_info.id().as_ref());
+    chain_length_to_block_ids.remove(chain_length_index)?;
 
-        println!(
-            "fetched {} intermediate blocks ({} per test), total distance {}",
-            blocks_fetched, blocks_per_test, total_distance
-        );
+    tips.remove(block_id)?;
 
-        assert!(blocks_per_test < 35.0);
+    if block_info.parent_id().as_ref() == root_id {
+        return Ok(RemoveTipResult::Done);
     }
 
-    #[test]
-    fn simultaneous_read_write() {
-        let mut rng = OsRng;
-        let store =
-            BlockStoreBuilder::file("file:test_simultaneous_read_write?mode=memory&cache=shared")
-                .build();
-
-        let mut conn = store.connect().unwrap();
-
-        let genesis_block = Block::genesis(None);
-        conn.put_block(&genesis_block).unwrap();
-        let mut blocks = vec![genesis_block];
+    let parent_permanent = permanent_store_index
+        .get(block_info.parent_id().as_ref())
+        .map(|maybe_block| maybe_block.is_some())?;
 
-        for _ in 1..SIMULTANEOUS_READ_WRITE_ITERS {
-            let last_block = blocks.get(rng.next_u32() as usize % blocks.len()).unwrap();
-            let block = last_block.make_child(None);
-            blocks.push(block.clone());
-            conn.put_block(&block).unwrap()
-        }
+    if parent_permanent {
+        return Ok(RemoveTipResult::HitPermanentStore {
+            id: block_info.parent_id().as_ref().to_vec(),
+        });
+    }
 
-        let mut conn_1 = store.connect().unwrap();
-        let blocks_1 = blocks.clone();
+    let parent_block_info_bin = info
+        .get(block_info.parent_id())?
+        .ok_or(ConsistencyFailure::MissingParentBlock)?;
+    let mut parent_block_info_reader: &[u8] = &parent_block_info_bin;
+    let mut parent_block_info = BlockInfo::deserialize(
+        &mut parent_block_info_reader,
+        id_size,
+        block_info.parent_id().clone(),
+    )?;
```

Comment:
Consider a helper function for getting block info from the sled store, this is becoming repetitive.